ENH: implement TorchImageToTorchImageFilter

Leengit opened the pull request.

Leengit commented:
This builds and tests clean on my local system.  The failures in the GitHub actions are unrelated to this pull request; instead they are the already known: failure to build pytorch on Micorosoft Windows & Apple OSX, failure to locate Torch_DIR when building ITKPyTorch on Ubuntu.

dzenanz commented on include/itkImageToTorchImageFilter.h:
```diff
@@ -19,60 +19,88 @@
  */
-template <typename TInputImage, typename TOutputImage>
-class ImageToTorchImageFilter : public ImageToImageFilter<TInputImage, TOutputImage>
+template< typename TInputImage, typename TOutputTorchImage >
```
The second parameter might not be needed. Even if kept, it should be defaulted to expected definition, e.g.:
```suggestion
template< typename TInputImage, typename TOutputTorchImage = TorchImage< typename TInputImage::PixelType, TInputImage::ImageDimension > >
```

Leengit replied:
I will omit the second parameter.

dzenanz commented on include/itkImageToTorchImageFilter.h:
```diff
@@ -19,60 +19,88 @@
 public:
-  ITK_DISALLOW_COPY_AND_ASSIGN(ImageToTorchImageFilter);
+  static constexpr bool ConceptCheck =
+    std::is_same< TInputImage, Image< typename TInputImage::PixelType, TInputImage::ImageDimension > >::value &&
```
This check seems tautological. In what case would it not be satisfied?

Leengit replied:
I want to verify that the supplied input image is an Image rather than a TorchImage.

dzenanz replied:
That didn't occur to me 😄

dzenanz commented on include/itkImageToTorchImageFilter.h:
```diff
@@ -19,60 +19,88 @@
-  ITK_DISALLOW_COPY_AND_ASSIGN(ImageToTorchImageFilter);
+  static constexpr bool ConceptCheck =
+    std::is_same< TInputImage, Image< typename TInputImage::PixelType, TInputImage::ImageDimension > >::value &&
+    std::is_same< TOutputTorchImage, TorchImage< typename TOutputTorchImage::PixelType, TOutputTorchImage::ImageDimension > >::value &&
```
```suggestion
std::is_same< TOutputTorchImage, TorchImage< typename TInputImage::PixelType, TInputImage::ImageDimension > >::value &&
```
With this change it is redundant with checks on lines 47 and 48. Without this change it is tautological too.

Leengit replied:
I will make it simpler, enabled by your suggestion to drop the second template parameter.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -18,48 +18,76 @@
+  // Get pointers to the input image.
+  typename Self::InputImageType *inputImage = const_cast< typename Self::InputImageType * >( this->GetInput( 0 ) );
+
+  if( inputImage != nullptr )
```
This should got into verify preconditions method.

Leengit replied:
Will do.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -18,48 +18,76 @@
+template< typename TInputImage, typename TOutputTorchImage >
+void
+ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
+::GenerateInputRequestedRegion()
```
You probably don't need to override this method.

Leengit replied:
I will move the logic from `GenerateInputRequestedRegion` to `VerifyPreconditions`.  I will no longer override the former.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -18,48 +18,76 @@
+  typename Self::OutputTorchImageType * const outputTorchImage = this->GetOutput();
+  const typename Self::OutputTorchImageSizeType outputSize = outputTorchImage->GetBufferedRegion().GetSize();
+
+  itkAssertOrThrowMacro( inputSize == outputSize, "ImageToTorchImageFilter: input and output images must have the same size in each dimension");
```
maybe add the word buffered region for clarification

Leengit replied:
Will do.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -18,48 +18,76 @@
+  itkAssertOrThrowMacro( inputSize == outputSize, "ImageToTorchImageFilter: input and output images must have the same size in each dimension");
+
+  // Do the memory copy of the whole image.
+  // Do we need to Allocate() the output?!!!
```
I think the output(s) are already allocated by the rest of the pipeline.

Leengit replied:
Good.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -18,48 +18,76 @@
-  {
-    out.Set(in.Get());
-  }
+  // We did the memory copy already.  Nothing to do here.
```
Then maybe go the route of overriding `GenerateData()`. Less method definitions. Mainly rename the current `BeforeThreadedGenerateData()` into `GenerateData()`. See what the superclass [GenerateData()](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.1.1/Modules/Core/Common/include/itkImageSource.hxx#L210-L245) does. I guess this class will only need GPU allocate and copy.

dzenanz commented on include/itkTorchImageToImageFilter.h:
```diff
@@ -19,60 +19,88 @@
-template <typename TInputImage, typename TOutputImage>
-class TorchImageToImageFilter : public ImageToImageFilter<TInputImage, TOutputImage>
+template< typename TInputTorchImage, typename TOutputImage >
+class ITK_TEMPLATE_EXPORT TorchImageToImageFilter : public ImageToImageFilter< TInputTorchImage, TOutputImage >
```
remove or default the output image type (similar to the other class)

Leengit replied:
Will do.

dzenanz commented on include/itkTorchImageToImageFilter.h:
```diff
@@ -19,60 +19,88 @@
+  static constexpr bool ConceptCheck =
+    std::is_same< TInputTorchImage, TorchImage< typename TInputTorchImage::PixelType, TInputTorchImage::ImageDimension > >::value &&
+    std::is_same< TOutputImage, Image< typename TOutputImage::PixelType, TOutputImage::ImageDimension > >::value &&
+    std::is_same< typename TInputTorchImage::PixelType, typename TOutputImage::PixelType >::value &&
```
lines 45 and 46 are redundant given lines 47 and 48.

Leengit replied:
Will do.

dzenanz commented on include/itkTorchImageToImageFilter.h:
```diff
@@ -19,60 +19,88 @@
+  void GenerateInputRequestedRegion() override;
 
-  using OutputRegionType = typename OutputImageType::RegionType;
+  void BeforeThreadedGenerateData() override;
```
Overriding `GenerateData` is simpler and sufficient.

Leengit replied:
Will do.

dzenanz commented on include/itkTorchImageToImageFilter.h:
```diff
@@ -19,60 +19,88 @@
+  void DynamicThreadedGenerateData( const OutputImageRegionType &outputRegion ) override;
 
 private:
 #ifdef ITK_USE_CONCEPT_CHECKING
```
Either remove this section or move the `#ifdef` to line 49.

Leengit replied:
Will do.

dzenanz commented on include/itkTorchImageToImageFilter.hxx:
```diff
@@ -18,48 +18,77 @@
+  // Get pointers to the input image.
+  typename Self::InputTorchImageType * const inputTorchImage = const_cast< typename Self::InputTorchImageType * >( this->GetInput( 0 ) );
+
+  if( inputTorchImage != nullptr )
```
Should go into [VerifyPreconditions()](https://itk.org/Doxygen/html/classitk_1_1ProcessObject.html#a73000f23064b7d2898fa21135080b4cf).

Leengit replied:
Will do.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -71,11 +72,18 @@ ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
-    torch::from_blob( ( void * )inputBufferPointer, outputTorchImage->m_Tensor.sizes(), outputTorchImage->m_Tensor.options().device( torch::kCPU ) );
-  outputTorchImage->m_Tensor = inputTensor.to( outputTorchImage->m_Tensor.options() );
+  const std::vector< int64_t > torchSize = outputTorchImage->ComputeTorchSize();
+  const c10::TensorOptions commonTensorOptions = torch::dtype( Self::OutputTorchImageType::TorchValueType ).layout( torch::kStrided );
```
I assume that `torch::kStrided` here means contiguous in memory?

Leengit replied:
Yes.

dzenanz commented on test/itkImageToTorchImageFilterTest.cxx:
```diff
@@ -224,6 +231,59 @@ int itkImageToTorchImageFilterTest(int argc, char *argv[])
+  using InputImageType = itk::Image< PixelType, ImageDimension >;
+  using OutputTorchImageType = itk::TorchImage< PixelType, ImageDimension >;
+
+  // Build an itkImage
```
This part of code might be replaced by [RandomImageSource](https://itk.org/ITKExamples/src/Core/TestKernel/GenerateRandomImage/Documentation.html).

Leengit replied:
That looks useful!  The next time that I write a test ....

dzenanz reviewed (commented):
This looks mostly good. Some in-line suggestions. I have not tested it.

dzenanz commented on include/itkImageToTorchImageFilter.h:
```diff
@@ -89,19 +89,15 @@ class ITK_TEMPLATE_EXPORT ImageToTorchImageFilter : public ImageToImageFilter< T
-  void BeforeThreadedGenerateData() override;
+  void GenerateData() override;
 
   void DynamicThreadedGenerateData( const typename Self::OutputTorchImageRegionType &outputRegion ) override;
```
This method is no longer needed.

Leengit replied:
Right you are!

Leengit replied:
Without DynamicThreadedGenerateData, the ctest fails with

> Description: itk::ERROR: itk::ERROR: ImageToTorchImageFilter(0x5647274e2070): Subclass should override this method!!! If old behavior is desired invoke this->DynamicMultiThreadingOff(); before Update() is called. The best place is in class constructor.

... so I have restored the override with the latest pushed commit.  Thoughts?

dzenanz replied:
Inside `GenerateData()` don't call [superclass' `GenerateData()`](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.1.1/Modules/Core/Common/include/itkImageSource.hxx#L210-L245). That's what is calling `DynamicThreadedGenerateData`. Instead, call `AllocateOutputs()` if you need it.

Leengit replied:
Thank you.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -53,23 +52,23 @@ ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
+  const typename Self::OutputTorchImageSizeType outputBufferedRegionSize = outputTorchImage->GetBufferedRegion().GetSize();
 
-  itkAssertOrThrowMacro( inputSize == outputSize, "ImageToTorchImageFilter: input and output images must have the same size in each dimension");
+  itkAssertOrThrowMacro( inputBufferedRegionSize == outputBufferedRegionSize, "ImageToTorchImageFilter: input and output images must have the same size in each dimension");
```
I meant the error message should mentioned buffered region, e.g.: _ImageToTorchImageFilter: input and output images' buffered regions must have the same size in each dimension_

Leengit replied:
Will do.

dzenanz commented on include/itkImageToTorchImageFilter.hxx:
```diff
@@ -87,9 +86,9 @@ ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
 void
-ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
+ImageToTorchImageFilter< TInputImage >
 ::DynamicThreadedGenerateData( const typename Self::OutputTorchImageRegionType &itkNotUsed( outputRegion ) )
```
This method can be removed now.

Leengit replied:
Will do.

dzenanz commented on include/itkTorchImageToImageFilter.hxx:
```diff
@@ -83,9 +82,9 @@ TorchImageToImageFilter< TInputTorchImage, TOutputImage >
 void
-TorchImageToImageFilter< TInputTorchImage, TOutputImage >
+TorchImageToImageFilter< TInputTorchImage >
 ::DynamicThreadedGenerateData( const typename Self::OutputImageRegionType &itkNotUsed( outputRegion ) )
```
Can be removed.

Leengit replied:
Will do.

dzenanz reviewed (approved):
Looks good. Minor in-line remarks. I prefer this to be squashed upon merging.

Commit:
ENH: Implement conversions between itk::TorchImage and itk::Image.

- Implement `TorchImageToTorchImageFilter`, `ImageToTorchImageFilter`,
  and `TorchImageToImageFilter`.
- Add tests.

thewtex commented on include/itkImageToTorchImageFilter.h:
```diff
@@ -0,0 +1,106 @@
+ */
+template< typename TInputImage >
+class ITK_TEMPLATE_EXPORT ImageToTorchImageFilter
+  : public ImageToImageFilter< TInputImage, TorchImage< typename TInputImage::PixelType, TInputImage::ImageDimension > >
```
We want the ability to use the filter, *Input, Output* with all the template arguments:

- *Image, TorchImage*
- *TorchImage, Image*
- *TorchImage, TorchImage*

without requiring a change to the class used. Also, we want to re-use the CPU image information metadata transformation logic, and only write one Torch implementation of the `GenerateData`. This implementation requires client code to support three different classes, requires re-implementation of the metadata transformation logic, and would require multiple copies of the `GenerateData` logic. The design pattern should follow that of `GPUImageToImageFilter`:

 https://github.com/InsightSoftwareConsortium/ITK/blob/4f6fc719bd1e51cbd8f12f15ef3286a18e006996/Modules/Core/GPUCommon/include/itkGPUImageToImageFilter.h#L26-L41

Commit:
Renaming and combining TorchImage base filter classes

Commit:
Deleting obsolete files.

dzenanz commented on include/itkTorchPixelHelper.h:
```diff
@@ -80,8 +80,8 @@ class TorchPixelHelper< TPixelType, typename std::enable_if< std::is_arithmetic<
-  static constexpr unsigned int NumberOfComponents = 1;
-  static constexpr unsigned int SizeOf = NumberOfComponents;
+  static constexpr unsigned int NumberOfTopLevelComponents = 1;
+  static constexpr unsigned int NumberOfDeepComponents = NumberOfTopLevelComponents;
```
```suggestion
  static constexpr unsigned int NumberOfDeepComponents = 1;
```

dzenanz reviewed (commented):
Looks good.

Leengit committed:
Update include/itkTorchPixelHelper.h

Co-authored-by: Dženan Zukić <dzenan.zukic@kitware.com>

Leengit commented:
Suggestions for To do:

1.  `TorchImage` should inherit from `Image` rather than `ImageBase`.  This enables `Image` functionality, such as a pixel buffer on the CPU, and is useful for  implicit conversion between `Image` instances in CPU memory and `TorchImage` instances in GPU or CPU memory.

1. Much as with `GPUImage` a mechanism needs to be set up to **lazily** keep the `TorchImage` and `Image` pixel data in sync.  If the CPU buffer isn't ever used then the laziness should extend to never allocating that buffer.  The synchronization might be achieved via classes that parallel `GPUDataManager` and `GPUImageDataManager`.

    1. The implementation for `GetPixel` and `SetPixel` in the current version of the `ITKPyTorch` module operate directly on the `TorchImage` pixel data, but whether this is correct should be considered in the context of the `GPUImage` data management approach.  These and other methods could instead operate on the `Image` buffer or throw exceptions.

    1. When adapting code from `GPUImage`, please change the word `dirty` to `stale`.  Within the `GPUImage` implementation, `dirty` is being used to mean that the associated pixel buffer is not up-to-date, but in [other contexts](https://en.wikipedia.org/wiki/Page_cache) `dirty` means the exact opposite: a buffer that is more up-to-date than its mirror and needs to be mirrored. 

1. The actual copying between CPU buffer and TorchImage's `torch::tensor` representation can be copied from commit 90371e8d9705006b491d713d06fd28ef4d33cd0b.  Look at the `GenerateData` methods in the (now deleted) files `itkImageToTorchImageFilterSave.hxx` and `itkTorchImageToImageFilterSave.hxx`.  Tests for the functionality of this code can be found in the same commit, in the (now deleted) files `itkImageToTorchImageFilterTestSave.cxx` and `itkTorchImageToImageFilterTestSave.cxx`.

Commit:
STYLE: clang formatting

Commit:
ENH: Use strings instead of numbers for versions so that 3.1 != 3.10
ENH: Update itk and pytorch versions

dzenanz commented:
@Leengit did you accidentally close this PR?

Leengit commented:
I deleted my Leengit fork of the repository.  I had no intention of
affecting the main repository, pull requests, etc.
